pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -757,9 +780,91 @@ func (gce *GCECloud) EnsureLoadBalancer(clusterName string, apiService *api.Serv
 
 	status := &api.LoadBalancerStatus{}
 	status.Ingress = []api.LoadBalancerIngress{{IP: ipAddress}}
+
 	return status, nil
 }
 
+func serviceNeedsHealthCheck(service *api.Service) (b bool) {
+	if l, ok := service.Annotations[apiservice.AnnotationHandleExternalTraffic]; ok {
+		if l == apiservice.AnnotationValueExternalEndpointsLocal {
+			return true
+		} else if l == apiservice.AnnotationValueExternalEndpointsGlobal {
+			return false
+		} else {
+			glog.Errorf("Invalid value for annotation %v", apiservice.AnnotationHandleExternalTraffic)
+			return false
+		}
+	}
+	return false
+}
+
+func getServiceHealthCheckNodePort(service *api.Service) (int32, error) {
+	if serviceNeedsHealthCheck(service) {
+		if l, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; ok {
+			p, err := strconv.Atoi(l)
+			if err != nil {
+				glog.Errorf("Failed to parse annotation %v: %v", apiservice.AnnotationHealthCheckNodePort, err)
+				return 0, fmt.Errorf("Failed to parse annotation %v: %v", apiservice.AnnotationHealthCheckNodePort, err)
+			}
+			return int32(p), nil
+		} else {
+			glog.Errorf("Service does not contain necessary annotation %v", apiservice.AnnotationHealthCheckNodePort)
+		}
+	}
+	return 0, fmt.Errorf("No health check needed")
+}
+
+/* Return the path programmed into the Cloud LB Health Check */
+func getServiceHealthCheckPathPort(service *api.Service) (string, int32) {
+	if !serviceNeedsHealthCheck(service) {
+		return "", 0
+	}
+	port, err := getServiceHealthCheckNodePort(service)
+	if err != nil {
+		return "", 0
+	}
+	// Format using the healthcheckparser library shared with kube-proxy parsing side.
+	return healthcheckparser.GenerateURL(service.Namespace, service.Name), port
+
+}
+
+func (gce *GCECloud) createHttpHealthCheckIfNeeded(name, path string, port int32) (hc *compute.HttpHealthCheck, err error) {
+	name = fmt.Sprintf("%v-%v", name, port)
+	// TODO: Trim name
+	if len(name) > nameLengthLimit {
+		return nil, fmt.Errorf("Cannot create health check, name too long %v(%d) only allowed %d chars", name, len(name), nameLengthLimit)
+	}
+	hc, err = gce.GetHttpHealthCheck(name)
+	if hc == nil || err != nil && isHTTPErrorCode(err, http.StatusNotFound) {
+		glog.Infof("Did not find health check %v, creating port %v path %v", name, port, path)
+		newHC := &compute.HttpHealthCheck{
+			Name:        name,
+			Port:        int64(port),
+			RequestPath: path,
+			Host:        "",
+			Description: "kubernetes HTTP health check for L4 loadbalancer",
```

Comment:
We should use the description the same way we do for other objects - a JSON-string of labels.